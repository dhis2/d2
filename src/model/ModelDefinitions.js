import { checkType, isType } from '../lib/check'

/**
 * Contains all the `ModelDefinition`s that are available. The definitions are properties on the object.
 * This would be used as a main entry point to do any interaction.
 *
 * After calling the initialise function `d2({baseUrl: 'dhis/api'})` this object is the `models` property
 * that allows you to access
 *
 * @example
 * models.dataElement.getList();
 *
 * @memberof module:model
 */
class ModelDefinitions {
    // TODO: Elaborate this documentation
    /**
     * This will allow you to add your own custom ModelDefinitions.
     *
     * The Definition object should have the following properties
     * `modelName, modelNamePlural, modelOptions, properties, validations`
     *
     * @param {ModelDefinition} modelDefinition Add a model definition to the definitions collection
     *
     * @example
     * models.add({name: 'MyDefinition', plural: 'MyDefinitions', endPointname: '/myDefinition'});
     */
    add(modelDefinition) {
        try {
            checkType(modelDefinition.name, 'string')
        } catch (e) {
            throw new Error(
                'Name should be set on the passed ModelDefinition to add one'
            )
        }

        if (this[modelDefinition.name]) {
            throw new Error(
                ['Model', modelDefinition.name, 'already exists'].join(' ')
            )
        }
        this[modelDefinition.name] = modelDefinition

        if (isType(modelDefinition.plural, 'string')) {
            this[modelDefinition.plural] = modelDefinition
        }
    }

    /**
     * Map through the modelDefinitions like you would with a simple `Array.map()`
     *
     * @param {Function} transformer Transformer function that will be run for each `ModelDefinition`
     * @returns {Array} Array with the `ModelDefinition` objects.
     *
     * @example
     * models.mapThroughDefinitions(definition => console.log(definition.name);
     *
     * @note {info} When mapping through the definition list `transformer` is called with the just the definition
     * Unlike other map functions, no index or the full object is being passed.
     *
     * @note {warn} The resulting array contains references to the actual objects. It does not work like immutable array functions.
     *
     */
    mapThroughDefinitions(transformer) {
        checkType(transformer, 'function', 'transformer')

        return Object.keys(this)
            .filter(
                modelDefinition =>
                    this.hasOwnProperty(modelDefinition) &&
                    !(this[modelDefinition].plural === modelDefinition)
            )
            .map(modelDefinition => transformer(this[modelDefinition]))
    }
}

// Model definitions singleton!
function getModelDefinitions() {
    if (getModelDefinitions.modelDefinitions) {
        return getModelDefinitions.modelDefinitions
    }
    return (getModelDefinitions.modelDefinitions = new ModelDefinitions())
}

ModelDefinitions.getModelDefinitions = getModelDefinitions

export default ModelDefinitions
