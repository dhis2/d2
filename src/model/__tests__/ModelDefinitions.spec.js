import ModelDefinitions from '../ModelDefinitions';

describe('D2 models', () => {
    let models;

    class ModelDefinition {
        constructor(name, plural) {
            this.name = name;
            this.plural = plural;
        }
    }

    beforeEach(() => {
        models = new ModelDefinitions();
    });

    it('should be an object', () => {
        expect(models).toBeInstanceOf(Object);
    });

    it('should not be allowed to be called without new', () => {
        expect(() => ModelDefinitions()).toThrowErrorMatchingSnapshot();
    });

    describe('add method', () => {
        let dataElementModelDefinition;

        beforeEach(() => {
            dataElementModelDefinition = new ModelDefinition('dataElement');
        });

        it('should be a function', () => {
            expect(models.add).toBeInstanceOf(Function);
        });

        it('should add a property to the models object', () => {
            models.add(dataElementModelDefinition);

            expect(models.dataElement).toBeInstanceOf(ModelDefinition);
        });

        it('should throw an error when trying to add something that already exists', () => {
            function shouldThrow() {
                models.add(dataElementModelDefinition);
            }
            models.add(dataElementModelDefinition);

            expect(shouldThrow).toThrowError('Model dataElement already exists');
        });

        it('should reject a ModelDefinition that does not have a name property', () => {
            function shouldThrow() {
                models.add({ apiEndPoint: '/dataElement' });
            }
            models.add(dataElementModelDefinition);

            expect(shouldThrow).toThrowError('Name should be set on the passed ModelDefinition to add one');
        });

        it('should add the plural version to the object', () => {
            const indicatorDefinition = new ModelDefinition('indicator', 'indicators');

            models.add(indicatorDefinition);

            expect(models.indicator).toBeInstanceOf(ModelDefinition);
            expect(models.indicator).toBe(models.indicators);
        });
    });

    describe('mapThroughDefinitions method', () => {
        beforeEach(() => {
            models.add({ name: 'dataElement' });
            models.add({ name: 'dataValue' });
            models.add({ name: 'user' });
            models.add({ name: 'userGroup' });
        });

        it('should should be a function', () => {
            expect(models.mapThroughDefinitions).toBeInstanceOf(Function);
        });

        it('should return an array of ModelDefinitions', () => {
            const expectedArray = [
                { name: 'dataElement' },
                { name: 'dataValue' },
                { name: 'user' },
                { name: 'userGroup' },
            ];
            function returnValue(item) {
                return item;
            }

            expect(models.mapThroughDefinitions(returnValue)).toEqual(expectedArray);
        });

        it('should throw if the transformer passed is not a function', () => {
            expect(() => models.mapThroughDefinitions('')).toThrowError('Expected transformer to have type function');
            expect(() => models.mapThroughDefinitions({})).toThrowError('Expected transformer to have type function');
        });

        it('should not map through properties that are the plural versions', () => {
            const iterator = jest.fn();

            models.add({ name: 'indicator', plural: 'indicators' });

            models.mapThroughDefinitions(iterator);

            expect(iterator).toHaveBeenCalledTimes(5);
        });
    });
});
