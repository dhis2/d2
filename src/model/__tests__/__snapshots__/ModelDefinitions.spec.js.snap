// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`D2 models should not be allowed to be called without new 1`] = `"Class constructor ModelDefinitions cannot be invoked without 'new'"`;
